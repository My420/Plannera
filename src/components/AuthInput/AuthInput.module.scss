@import '../../sass/variables.scss';

.label {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  width: 300px;
}

.input {
  box-sizing: border-box;

  width: 300px;
  height: 40px;

  padding: 0 5px;

  border: none;
  border-radius: 3px;

  font-size: 14px;
  line-height: 16px;
  text-align: center;
  color: $fonts-auth-color;
}

.error,
.valid {
  min-height: 14px;

  padding: 10px 2px 0 2px;
  font-size: 12px;
  line-height: 14px;
  color: $fonts-app-error-color;
}

.valid {
  opacity: 0;
}

.notValid {
  background-color: red;
}
