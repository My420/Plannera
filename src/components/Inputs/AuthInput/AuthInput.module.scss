@import '../../../sass/variables.scss';
@import '../../../sass/mixins.scss';

.label {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  width: 300px;
}

.input {
  @include auth-input;
  box-sizing: border-box;

  width: 300px;
  height: 40px;

  padding: 0 5px;

  border: none;
  border-radius: 3px;

  font-size: 14px;
  line-height: 16px;
  text-align: center;
  color: $fonts-auth-color;
}

.error,
.valid {
  min-height: 14px;

  padding: 10px 2px 0 2px;
  font-size: 12px;
  font-weight: 700;
  line-height: 14px;
  color: $fonts-app-error-color;
}

.valid {
  opacity: 0;
}

.notValid {
  background-color: red;
}

@media (max-width: $mobile-large) {
  .label {
    width: 280px;
  }

  .input {
    width: 280px;
    height: 30px;
  }

  .error,
  .valid {
    min-height: 11px;
    font-size: 11px;
    line-height: 11px;
  }
}
