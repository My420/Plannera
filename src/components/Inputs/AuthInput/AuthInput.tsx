import React from 'react';
import { InputAttributeType } from '../../../types/types';
import styles from './AuthInput.module.scss';

export interface AuthInputProps extends React.HTMLAttributes<HTMLInputElement> {
  label: string;
  type: InputAttributeType;
  name: string;
  id: string;
  placeholder: string;
  extensionContainerClass?: string;
  extensionInputClass?: string;
  value: string;
  isValid: boolean;
  error: string;
  onInputChange: (name: string, value: string) => void;
}

const AuthInput: React.FC<AuthInputProps> = ({
  value,
  onInputChange,
  label,
  name,
  id,
  type,
  placeholder,
  extensionContainerClass,
  extensionInputClass,
  isValid,
  error,
  ...rest
}) => {
  const handleChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
    const { name: inputName, value: newValue } = evt.currentTarget;
    onInputChange(inputName, newValue);
  };

  return (
    <label htmlFor={id} className={`${styles.label} ${extensionContainerClass}`}>
      <span className="visually-hidden">{label}</span>
      <span className={`${isValid ? styles.valid : styles.error}`}>
        {error}
        {' '}
      </span>
      <input
        value={value}
        id={id}
        className={`${styles.input} ${extensionInputClass}`}
        name={name}
        type={type}
        placeholder={placeholder}
        onChange={handleChange}
        {...rest}
      />
    </label>
  );
};

AuthInput.defaultProps = {
  extensionContainerClass: '',
  extensionInputClass: '',
};

export default AuthInput;
