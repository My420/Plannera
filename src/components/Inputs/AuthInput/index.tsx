import React from 'react';
import AuthInput, { AuthInputProps } from './AuthInput';

const areEqual = (prevProps: AuthInputProps, nextProps: AuthInputProps) => {
  const { value: prevValue, isValid: prevIsValid, error: prevError } = prevProps;
  const { value: nextValue, isValid: nextIsValid, error: nextError } = nextProps;
  return prevValue === nextValue && prevIsValid === nextIsValid && prevError === nextError;
};

export default React.memo(AuthInput, areEqual);
