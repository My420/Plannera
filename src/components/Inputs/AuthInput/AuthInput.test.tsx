import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AuthInput, { AuthInputProps } from './AuthInput';

describe('test AuthInput component', () => {
  const mockFn = jest.fn();
  const props: AuthInputProps = {
    label: 'test label',
    type: 'text',
    name: 'test name',
    id: 'testID',
    value: '',
    placeholder: 'enter test text',
    isValid: false,
    error: '',
    onInputChange: mockFn,
  };

  afterEach(cleanup);

  test('should match snapshot', () => {
    const { asFragment } = render(<AuthInput {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render label that associate with input', () => {
    render(<AuthInput {...props} />);
    const elem = screen.getByLabelText(props.label);
    expect(elem.tagName).toBe('INPUT');
  });

  test("should render tag span with label text and 'visually-hidden' class", () => {
    render(<AuthInput {...props} />);
    const elem = screen.getByText(props.label);
    expect(elem).toHaveClass('visually-hidden');
  });

  test('should render tag input with correct attr', () => {
    render(<AuthInput {...props} />);
    const elem = screen.getByPlaceholderText(props.placeholder);
    expect(elem.tagName).toBe('INPUT');
    expect(elem).toHaveValue(props.value);
    expect(elem).toHaveAttribute('type', props.type);
  });

  test('should render correct tag span with error if props.isValid = false', () => {
    render(<AuthInput {...props} />);
    const elem = screen.getByTestId('error');
    expect(elem).toHaveClass('error');
    expect(elem).toHaveTextContent(props.error);
  });

  test('should render correct tag span with error if props.isValid = true', () => {
    const errorText = 'error error';
    const newProps = { ...props, isValid: true, error: errorText };
    render(<AuthInput {...newProps} />);
    const elem = screen.getByText(errorText);
    expect(elem).toHaveClass('valid');
  });

  test('when user enters text should call props.onInputChange function', async () => {
    render(<AuthInput {...props} />);
    const elem = screen.getByPlaceholderText(props.placeholder);
    const text = 'test user input';
    await userEvent.type(elem, text, { allAtOnce: true });
    expect(mockFn).toHaveBeenCalledWith(props.name, text);
  });
});
