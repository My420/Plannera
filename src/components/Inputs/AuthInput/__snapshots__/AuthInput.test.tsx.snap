// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AuthInput component should match snapshot 1`] = `
<DocumentFragment>
  <label
    class="label "
    for="testID"
  >
    <span
      class="visually-hidden"
    >
      test label
    </span>
    <span
      class="error"
      data-testid="error"
    />
    <input
      class="input "
      id="testID"
      name="test name"
      placeholder="enter test text"
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;
